### ISSUE TITLE
Investigate Performance of Math Functions in .NET Standard 2.1+

### ISSUE THREAD

**Issue by NightOwl888** [apache/lucenenet#297]
As part of our effort to improve performance (#261), Microsoft has [added support](https://github.com/dotnet/runtime/issues/20322) for some of the functions we use in [MathUtil](https://github.com/apache/lucenenet/blob/a7f7c40895b156681beaea22e1da8f46e265a98c/src/Lucene.Net/Util/MathUtil.cs).

```c#
    // Hyperbolic Functions
    public static double Acosh(double);                 // Compute area hyperbolic cosine
    public static double Asinh(double);                 // Compute area hyperbolic sine
    public static double Atanh(double);                 // Compute area hyperbolic tangent

    public static double Log(double, double);
```

We should run benchmarks to see if the built-in implementation is faster than ours, and if so reference the built-in `Math` class inside of `MathUtil` in .NET Standard 2.1+ as well as consider the same optimizations to `MathUtil` for other target frameworks.

It also looks like `public static int Log(long, long)` may be able to be optimized.

**Comment by jeme** [comment]
Some very early indications:

```
// * Summary *

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18363.1016 (1909/November2018Update/19H2)
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.100-preview.7.20366.15
  [Host]     : .NET Core 3.1.6 (CoreCLR 4.700.20.26901, CoreFX 4.700.20.31603), X64 RyuJIT
  DefaultJob : .NET Core 3.1.6 (CoreCLR 4.700.20.26901, CoreFX 4.700.20.31603), X64 RyuJIT


|         Method |     Mean |    Error |   StdDev |
|--------------- |---------:|---------:|---------:|
| MathUtil_Acosh | 115.0 us |  2.26 us |  3.58 us |
|     Math_Acosh | 243.6 us |  4.74 us |  6.94 us |
| MathUtil_Asinh | 111.5 us |  1.55 us |  1.45 us |
|     Math_Asinh | 306.2 us |  3.28 us |  2.91 us |
|   MathUtil_Log | 139.7 us |  2.79 us |  4.42 us |
|       Math_Log | 180.5 us |  2.39 us |  1.99 us |
| MathUtil_Atanh | 649.9 us | 10.32 us |  8.06 us |
|     Math_Atanh | 519.9 us | 10.33 us | 11.48 us |
```

**Comment by NightOwl888** [comment]
@jeme 

Thanks for the info. These results are similar to the benchmark that I did earlier. I find the results a bit surprising since Microsoft implemented these functions in C++, but apparently ours are more efficient except for `Atanh`.

When I posted this task, I was working under the assumption that these methods were being used by other Lucene.NET components, thus having a potential impact on performance. But according to code lens, the only methods that are referenced (other than in tests) are `Log(long, int)` and `Gcd(long, long)`, neither of which have a counterpart in .NET. 

`Gcd(long, long)` is the only one that actually has an impact in the latest codecs and it has already been modified to use J2N's version of `TrailingZeroCount()` which will call the hardware intrinsic method on .NET Core 3.x and Windows if we add a .NET Core 3.x target framework to Lucene.NET.

That being said, since there doesn't appear to be anything impactful we can improve on `MathUtil` itself, we should focus our optimization efforts elsewhere.
